### ISSUE TITLE
Slower matrixmultiply in nightly-2017-12-16

### ISSUE THREAD

**Issue by bluss** [rust-lang/rust#46923]
- Good rustc 1.24.0-nightly (0077d128d 2017-12-14)  (rustup `nightly-2017-12-15`)
- Bad rustc 1.24.0-nightly (77efd6800 2017-12-15) (rustup `nightly-2017-12-16`)

[matrixmultiply](https://github.com/bluss/matrixmultiply) version 0.1.14. This code is deliberately seeking autovectorization, and it's understood that such things could be fragile; but for a long time it has worked fine.

Before:

```
RUSTFLAGS=-Ctarget-feature=+avx cargo +nightly-2017-12-15 bench -v m127

test mat_mul_f32::m127             ... bench:     121,631 ns/iter (+/- 10,097)
test mat_mul_f64::m127             ... bench:     198,662 ns/iter (+/- 13,402)
```

After:

```
RUSTFLAGS=-Ctarget-feature=+avx cargo +nightly-2017-12-16 bench -v m127

test mat_mul_f32::m127             ... bench:     209,268 ns/iter (+/- 12,809)
test mat_mul_f64::m127             ... bench:     295,810 ns/iter (+/- 2,167)
```

The PRs in that range are:

`git log --oneline --first-parent 0077d128d..77efd6800`
77efd6800c Auto merge of #46623 - eddyb:issue-46449, r=nagisa
50f6c3ece0 Auto merge of #46745 - steveklabnik:rollup, r=steveklabnik
5a0dc2d06d Auto merge of #46740 - oli-obk:rls, r=nrc
b4f323254a Auto merge of #46706 - sunjay:gat-lifetimes, r=nikomatsakis
04b23449c0 Auto merge of #46653 - estebank:str-as-ch, r=petrochenkov
84feab34e4 Auto merge of #46537 - pnkfelix:two-phase-borrows, r=arielb1
9331031909 Auto merge of #46701 - eddyb:vector-newtypes, r=nagisa
2f351415e5 Auto merge of #46641 - petrochenkov:nohelp2, r=nikomatsakis

**Comment by pmarcelll** [comment]
This might be the same issue as #46897.

**Comment by eddyb** [comment]
@pmarcelll I think so too. @bluss As per https://github.com/rust-lang/rust/issues/46897#issuecomment-353577250, can you retry with the latest nightly?

**Comment by bluss** [comment]
New results (no fix ☹️ )

- Bad rustc 1.24.0-nightly (7eb64b86c 2017-12-20)
- Bad rustc 1.24.0-nightly (250b49205 2017-12-21)
- Bad rustc 1.24.0-nightly (5165ee9e2 2017-12-22)

**Comment by bluss** [comment]
I haven't had time to investigate, but I'll do that over christmas.

**Comment by arielb1** [comment]
This really looks like #46897

**Comment by bluss** [comment]
Probably. The reason I discovered it was due to that issue, but when I read it it seemed to talk about a different nightly version as the breaking point, so I reported this too.

**Comment by bluss** [comment]
Fixed in rustc 1.24.0-nightly (1abeb436d 2017-12-27)   🌟 @eddyb 🌟 

```
RUSTFLAGS=-Ctarget-feature=+avx\ -Ccodegen-units=1 cargo +nightly bench -v m127
test mat_mul_f32::m127             ... bench:     123,960 ns/iter (+/- 8,354)
test mat_mul_f64::m127             ... bench:     194,921 ns/iter (+/- 3,862)
```

So it was resolved like #46897
